util/oidc/oidc.go

```diff
@@ -336,6 +346,41 @@ func (a *ClientApp) HandleLogin(w http.ResponseWriter, r *http.Request) {
 	http.Redirect(w, r, url, http.StatusSeeOther)
 }
 
+// getAzureKubernetesServiceAccountToken returns the specified file's content, which is expected to be a Kubernetes service account token.
+// Kubernetes is responsible for updating the file as service account tokens expire.
+func (a *ClientApp) getAzureKubernetesServiceAccountToken(context.Context) (string, error) {
+	file := ""
+	ok := false
+	if file == "" {
+		if file, ok = os.LookupEnv("AZURE_FEDERATED_TOKEN_FILE"); !ok {
+			return "", errors.New("no token file specified. Check pod configuration or set TokenFilePath in the options")
+		}
+	}
+
+	a.mtx.RLock()
+	if a.expires.Before(time.Now()) {
+		// ensure only one goroutine at a time updates the assertion
+		a.mtx.RUnlock()
+		a.mtx.Lock()
+		defer a.mtx.Unlock()
+		// double check because another goroutine may have acquired the write lock first and done the update
+		if now := time.Now(); a.expires.Before(now) {
+			content, err := os.ReadFile(file)
+			if err != nil {
+				return "", err
+			}
+			a.assertion = string(content)
+			// Kubernetes rotates service account tokens when they reach 80% of their total TTL. The shortest TTL
+			// is 1 hour. That implies the token we just read is valid for at least 12 minutes (20% of 1 hour),
+			// but we add some margin for safety.
+			a.expires = now.Add(10 * time.Minute)
```

Comment:
You will need tests to validate the race conditions and expiration logic. You will need to define a struct that you on which you can configure the expiration time.

Reply:
It can be done without creating struct as I can mark the expiration time in the test itself to mimic the behaviour. I wanted to test the number of time file is read but it is too complicated and will require a lot of refactoring. Will skip that part of the test.